README.md
# variant

## Synopsis

Topic tools for treating ROS messages as type-erased variants.

**Author(s):** Ralf Kaestner

**Maintainer:** Ralf Kaestner <ralf.kaestner@gmail.com>

**Licsense:** GNU Lesser General Public License (LGPL)

**Operating system(s):** Debian-based Linux, Mac OS X

**Package PPA:** Not available

## Description

This project provides a partially templated C++ API for treating ROS messages
as type-erased variants. Similar to the Python API of ROS, it thus renders ROS
messages and their members accessible without the requirement for including any
message headers at compile time. The goal of this functionality is to encourage
developers to contribute more C++ packages which are intended to inspect and
operate on any message type.

This project further provides a test package which practically demonstrates
API usage by example.

## Installation

### Installing from packages (recommended for Ubuntu LTS users)

The maintainers of this project do not yet provide binary packages.

### Building from source

This project may be built from source using catkin.

Here, we assume you intend to build the project for the ROS distribution
named `ROS_DISTRO`.

#### Installing build dependencies

The build dependencies of this project are available from the standard
package repositories of recent Ubuntu and ROS releases. To install them,
simply use the command

```shell
sudo apt-get install ros-ROS_DISTRO-roscpp, ros-ROS_DISTRO-std-msgs
```

#### Building with catkin (recommended for ROS developers)

Assuming that you have cloned the project sources into `PROJECT_DIR`, you
may attempt to build this project using catkin as follows:

* Create the directory structure for your catkin workspace by issuing

  ```shell
  mkdir -p CATKIN_WS_DIR/src
  ```

* Assuming that your ROS environment has been set up properly, initialize the
  catkin workspace using the command

  ```shell
  catkin_init_workspace CATKIN_WS_DIR/src
  ```

* Individually link the directories containing the project's package source
  trees into your catkin workspace source directory through

  ```shell
  ln -sf PROJECT_DIR/roscpp_nodewrap* CATKIN_WS_DIR/src
  ```

* Switch into the catkin workspace directory by 

  ```shell
  cd CATKIN_WS_DIR
  ```

* In the catkin workspace directory, run 

  ```shell
  catkin_make
  ```

  to start the build

## API documentation

This project does not yet build its API documentation. Some basic documentation
is however inlined with the interface definitions and can thus be found in the
source headers.

## Feature requests and bug reports

If you would like to propose a feature for this project, please consider
contributing or send a feature request to the project authors. Bugs may be
reported through the project's issue page.

## Further reading

For additional information of the Robot Operating System (ROS), please refer
to the official [ROS documentation](http://wiki.ros.org).

## Build Status

### Devel Job Status

| | Indigo  |
| --- | --- |
| variant | [![Build Status](http://build.ros.org/buildStatus/icon?job=Idev__variant__ubuntu_trusty_amd64)](http://build.ros.org/job/Idev__variant__ubuntu_trusty_amd64/) |

### Release Job Status

| | Indigo |
| --- | --- |
| variant | [![Build Status](http://build.ros.org/buildStatus/icon?job=Ibin_uT64__variant__ubuntu_trusty_amd64__binary)](http://build.ros.org/job/Ibin_uT64__variant__ubuntu_trusty_amd64__binary/) |
| variant_msgs | [![Build Status](http://build.ros.org/buildStatus/icon?job=Ibin_uT64__variant_msgs__ubuntu_trusty_amd64__binary)](http://build.ros.org/job/Ibin_uT64__variant_msgs__ubuntu_trusty_amd64__binary/) |
| variant_topic_test | [![Build Status](http://build.ros.org/buildStatus/icon?job=Ibin_uT64__variant_topic_test__ubuntu_trusty_amd64__binary)](http://build.ros.org/job/Ibin_uT64__variant_topic_test__ubuntu_trusty_amd64__binary/) |
| variant_topic_tools | [![Build Status](http://build.ros.org/buildStatus/icon?job=Ibin_uT64__variant_topic_tools__ubuntu_trusty_amd64__binary)](http://build.ros.org/job/Ibin_uT64__variant_topic_tools__ubuntu_trusty_amd64__binary/) |


changelog
variant (0.1-29) unstable; urgency=low

  * Fixes related to OS X clang compatibility

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Wed, 10 Feb 2016 07:01:00 +0200

variant (0.1-28) unstable; urgency=low

  * Fixed some issues of subscriber

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Wed, 09 Dec 2015 14:53:00 +0200

variant (0.1-27) unstable; urgency=low

  * Fixed bug in MessageDefinition::setMessageType() which was caused by a
    wrong order of defining the required message types of a message type

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Fri, 20 Nov 2015 08:30:00 +0200

variant (0.1-26) unstable; urgency=low

  * Added message data type member access by name

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Tue, 10 Nov 2015 08:39:00 +0200

variant (0.1-25) unstable; urgency=low

  * Added type name resolution for member definitions neglecting package names
  * Added receipt time to variant subscriber callback

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 09 Nov 2015 15:18:00 +0200

variant (0.1-24) unstable; urgency=low

  * Code cleanups
  * Added variant publisher and subscriber
  * Added MD5 sum calculation for variant message types

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 30 Oct 2015 13:27:00 +0200

variant (0.1-23) unstable; urgency=low

  * Updated README

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 29 Oct 2015 13:09:00 +0200

variant (0.1-22) unstable; urgency=low

  * Added convenience serialization methods to message class

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 29 Oct 2015 13:05:00 +0200

variant (0.1-21) unstable; urgency=low

  * Added method for serializer creation to variants

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 29 Oct 2015 09:41:00 +0200

variant (0.1-20) unstable; urgency=low

  * Basic serializer testing

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 29 Oct 2015 07:34:00 +0200

variant (0.1-19) unstable; urgency=low

  * Added numeric value access for built-in types

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Wed, 28 Oct 2015 07:03:00 +0200

variant (0.1-18) unstable; urgency=low

  * Adapted README

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Tue, 27 Oct 2015 16:26:00 +0200

variant (0.1-17) unstable; urgency=low

  * Message variant tests successful

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Tue, 27 Oct 2015 16:13:00 +0200

variant (0.1-16) unstable; urgency=low

  * Message data type now separates constant and variable members

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 19 Oct 2015 11:10:52 +0200

variant (0.1-15) unstable; urgency=low

  * Reverted templated variant value to represent actual value type

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 19 Oct 2015 09:54:51 +0200

variant (0.1-14) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 19 Oct 2015 07:41:39 +0200

variant (0.1-13) unstable; urgency=low

  * Another progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Fri, 16 Oct 2015 15:45:00 +0200

variant (0.1-12) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Fri, 16 Oct 2015 13:03:12 +0200

variant (0.1-11) unstable; urgency=low

  * Refined traits, code cleanups

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 08 Oct 2015 14:02:55 +0200

variant (0.1-10) unstable; urgency=low

  * Removed SharedVariant and friend-relationship between Variant and
    CollectionVariant

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 05 Oct 2015 16:03:59 +0200

variant (0.1-9) unstable; urgency=low

  * Renamed some classes, code cleanups

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 05 Oct 2015 07:51:54 +0200

variant (0.1-8) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 01 Oct 2015 10:16:04 +0200

variant (0.1-7) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 20 Aug 2015 17:59:39 +0200

variant (0.1-6) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 17 Aug 2015 07:05:01 +0200

variant (0.1-5) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Wed, 12 Aug 2015 12:56:26 +0200

variant (0.1-4) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Mon, 10 Aug 2015 07:54:58 +0200

variant (0.1-3) unstable; urgency=low

  * Created test package
  * Moved test binaries to test package

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 06 Aug 2015 09:57:09 +0200

variant (0.1-2) unstable; urgency=low

  * Progress commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 06 Aug 2015 06:47:21 +0200

variant (0.1-1) unstable; urgency=low

  * Initial commit

 -- Ralf Kaestner <ralf.kaestner@gmail.com>  Thu, 30 Jul 2015 05:48:09 +0200


variant_topic_tools/src/MD5Sum.cpp
/******************************************************************************
 * Copyright (C) 2014 by Ralf Kaestner                                        *
 * ralf.kaestner@gmail.com                                                    *
 *                                                                            *
 * This program is free software; you can redistribute it and/or modify       *
 * it under the terms of the Lesser GNU General Public License as published by*
 * the Free Software Foundation; either version 3 of the License, or          *
 * (at your option) any later version.                                        *
 *                                                                            *
 * This program is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the               *
 * Lesser GNU General Public License for more details.                        *
 *                                                                            *
 * You should have received a copy of the Lesser GNU General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 ******************************************************************************/

#include "variant_topic_tools/MD5Sum.h"

namespace variant_topic_tools {

/*****************************************************************************/
/* Static initializations                                                    */
/*****************************************************************************/

const MD5Sum::Digest MD5Sum::initialDigest = {
  0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476
};

const boost::array<uint32_t, 64> MD5Sum::numBitRotations = {
  7, 12, 17, 22,  7, 12, 17, 22,  7, 12, 17, 22,  7, 12, 17, 22,
  5,  9, 14, 20,  5,  9, 14, 20,  5,  9, 14, 20,  5,  9, 14, 20,
  4, 11, 16, 23,  4, 11, 16, 23,  4, 11, 16, 23,  4, 11, 16, 23,
  6, 10, 15, 21,  6, 10, 15, 21,  6, 10, 15, 21,  6, 10, 15, 21
};

const boost::array<uint32_t, 64> MD5Sum::constants = {
  0xd76aa478, 0xe8c7b756, 0x242070db, 0xc1bdceee,
  0xf57c0faf, 0x4787c62a, 0xa8304613, 0xfd469501,
  0x698098d8, 0x8b44f7af, 0xffff5bb1, 0x895cd7be,
  0x6b901122, 0xfd987193, 0xa679438e, 0x49b40821,
  0xf61e2562, 0xc040b340, 0x265e5a51, 0xe9b6c7aa,
  0xd62f105d, 0x02441453, 0xd8a1e681, 0xe7d3fbc8,
  0x21e1cde6, 0xc33707d6, 0xf4d50d87, 0x455a14ed,
  0xa9e3e905, 0xfcefa3f8, 0x676f02d9, 0x8d2a4c8a,
  0xfffa3942, 0x8771f681, 0x6d9d6122, 0xfde5380c,
  0xa4beea44, 0x4bdecfa9, 0xf6bb4b60, 0xbebfbc70,
  0x289b7ec6, 0xeaa127fa, 0xd4ef3085, 0x04881d05,
  0xd9d4d039, 0xe6db99e5, 0x1fa27cf8, 0xc4ac5665,
  0xf4292244, 0x432aff97, 0xab9423a7, 0xfc93a039,
  0x655b59c3, 0x8f0ccc92, 0xffeff47d, 0x85845dd1,
  0x6fa87e4f, 0xfe2ce6e0, 0xa3014314, 0x4e0811a1,
  0xf7537e82, 0xbd3af235, 0x2ad7d2bb, 0xeb86d391
};

const MD5Sum::Block MD5Sum::padding = {
  0x00000080, 0x00000000, 0x00000000, 0x00000000,
  0x00000000, 0x00000000, 0x00000000, 0x00000000,
  0x00000000, 0x00000000, 0x00000000, 0x00000000,
  0x00000000, 0x00000000, 0x00000000, 0x00000000
};

/*****************************************************************************/
/* Constructors and Destructor                                               */
/*****************************************************************************/

MD5Sum::MD5Sum(const std::string& text) {
  clear();
  update(text);
}

MD5Sum::MD5Sum(const MD5Sum& src) :
  digest(src.digest),
  numDigestedBits(src.numDigestedBits),
  buffer(src.buffer) {
}

MD5Sum::~MD5Sum() {
}

/*****************************************************************************/
/* Accessors                                                                 */
/*****************************************************************************/

size_t MD5Sum::getNumDigestedBits() const {
  const uint8_t* numDigestedBits = reinterpret_cast<const uint8_t*>(
    this->numDigestedBits.data());
  size_t i = this->numDigestedBits.size()*sizeof(uint32_t)-sizeof(size_t);
  
  return *reinterpret_cast<const size_t*>(&numDigestedBits[i]);
}

MD5Sum::Digest MD5Sum::getDigest() const {
  Digest digest(this->digest);
  Size numDigestedBits(this->numDigestedBits);
  Block buffer(this->buffer);
  
  finalize(digest, numDigestedBits, buffer);
  
  return digest;
}

/*****************************************************************************/
/* Methods                                                                   */
/*****************************************************************************/

void MD5Sum::update(const std::string& text) {
  if (!text.empty())
    update(reinterpret_cast<const uint8_t*>(text.data()), text.length());
}

void MD5Sum::update(const uint8_t* bytes, size_t numBytes) {
  update(digest, numDigestedBits, buffer, bytes, numBytes);
}

void MD5Sum::clear() {
  digest = initialDigest;
  
  numDigestedBits.fill(0);
  buffer.fill(0);
}

void MD5Sum::write(std::ostream& stream) const {
  stream << toString();
}

std::string MD5Sum::toString() const {
  Digest digest = getDigest();
  
  char stringData[2*digest.size()*sizeof(uint32_t)+1];
  const uint8_t* digestData = reinterpret_cast<const uint8_t*>(
    digest.data());
  
  for (size_t i = 0; i < digest.size()*sizeof(uint32_t); ++i)
    sprintf(&stringData[2*i], "%02x", digestData[i]);
  stringData[2*digest.size()*sizeof(uint32_t)] = 0;
 
  return std::string(stringData);
}

void MD5Sum::transform(const Block& block, Digest& digest) {
  uint32_t a = digest[0];
  uint32_t b = digest[1];
  uint32_t c = digest[2];
  uint32_t d = digest[3];
  
  for (size_t i = 0; i < 64; ++i) {
    uint32_t f, g, h;
    
    if (i < 16) {
      f = (b & c) | (~b & d);
      g = i;
    }
    else if (i < 32) {
      f = (b & d) | (c & ~d);
      g = (5*i+1) % 16;
    }
    else if (i < 48) {
      f = b ^ c ^ d;
      g = (3*i+5) % 16;
    }
    else if (i < 64) {
      f = c ^ (b | ~d);
      g = (7*i) % 16;
    }

    h = d;
    d = c;
    c = b;
    b += rotateLeft(a+f+constants[i]+block[g], numBitRotations[i]);
    a = h;
  }
  
  digest[0] += a;
  digest[1] += b;
  digest[2] += c;
  digest[3] += d;
}

uint32_t MD5Sum::rotateLeft(uint32_t value, size_t numBits) {
  return (value << numBits)|(value >> (sizeof(uint32_t)*8-numBits));
}

void MD5Sum::update(Digest& digest, Size& numDigestedBits, Block& buffer,
    const uint8_t* bytes, size_t numBytes) {
  size_t i = numDigestedBits[0]/8 % 64;
  size_t lengthOfFirstPart = 64-i;
  size_t j = 0;
 
  if ((numDigestedBits[0] += (numBytes << 3)) < (numBytes << 3))
    ++numDigestedBits[1];
  numDigestedBits[1] += (numBytes >> 29);
 
  if (numBytes >= lengthOfFirstPart) {
    std::copy(&bytes[0], &bytes[lengthOfFirstPart], 
    &reinterpret_cast<uint8_t*>(buffer.data())[i]);
    transform(buffer, digest);
 
    for (j = lengthOfFirstPart; j+64 <= numBytes; j += 64) {
      Block block;

      std::copy(&bytes[j], &bytes[j+64], reinterpret_cast<uint8_t*>(
        block.data()));
      transform(block, digest);
    }
 
    i = 0;
  }
  else
    j = 0;
 
  std::copy(&bytes[j], &bytes[numBytes],
    &reinterpret_cast<uint8_t*>(buffer.data())[i]);
}

void MD5Sum::finalize(Digest& digest, Size& numDigestedBits, Block& buffer) {
  Size finalNumDigestedBits(numDigestedBits);
  size_t i = numDigestedBits[0]/8 % 64;
  size_t paddedLength = (i < 56) ? 56-i : 120-i;
  
  update(digest, numDigestedBits, buffer, reinterpret_cast<const uint8_t*>(
    padding.data()), paddedLength);
  update(digest, numDigestedBits, buffer, reinterpret_cast<const uint8_t*>(
    finalNumDigestedBits.data()), 8);
}

/*****************************************************************************/
/* Operators                                                                 */
/*****************************************************************************/

MD5Sum& MD5Sum::operator<<(const std::string& text) {
  update(text);
  return *this;
}

std::ostream& operator<<(std::ostream& stream, const MD5Sum& md5Sum) {
  md5Sum.write(stream);
  return stream;
}

}
